server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java

```diff
@@ -78,6 +85,48 @@ public void register(String aggregationName, List<ValuesSourceType> valuesSource
             }
         }
 
+        /**
+         * Register a new key generation function for the
+         * {@link org.elasticsearch.search.aggregations.bucket.composite.CompositeAggregation}.
+         * @param supplierClass the subclass of {@link CompositeSupplier} associated with the {@link CompositeValuesSourceBuilder} type this
+         *                      mapping is being registered for
+         * @param valuesSourceType the {@link ValuesSourceType} this mapping applies to
+         * @param compositeSupplier A function returning an appropriate
+         *                          {@link org.elasticsearch.search.aggregations.bucket.composite.CompositeValuesSourceConfig}
+         */
+        public void registerComposite(
+            Class<? extends CompositeSupplier> supplierClass,
+            ValuesSourceType valuesSourceType,
+            CompositeSupplier compositeSupplier
+        ) {
+            // TODO: Assert is almost definitely wrong here, but I don't know what the right thing to throw is.
+            assert compositeSupplier.getClass() == supplierClass;
```

Comment:
@nik9000 - What do you think is the right thing to throw from here?  This would be a startup-time error to plugin authors indicating that they loaded a miss-matched supplier.  Thinking about it, throwing from registration is pretty scary. If the plugin doesn't catch this, we could utterly break its whole registration process.

Reply:
```
public <T extend CompositeSupplier> registerComposite(Class<T> supplierClass, ValuesSourceType vst, T compositeSupplier)
```
would make it fail at compile time. Unless the caller does mean things with type erasure. And if they do that then the deserve all the terrible failure they'll get.